pincer/utils/event_mgr.py

```diff
@@ -0,0 +1,287 @@
+    ----------
+    event_name : str
+        The name of the event.
+    check : Optional[Callable[[Any], bool]]
+        ``can_be_set`` only returns true if this function returns true.
+        Will be ignored if set to None.
+
+    Attributes
+    ----------
+    event : :class:`asyncio.Event`
+        Even that is used to wait until the next valid discord event.
+    return_value : Optional[str]
+        Used to store the arguments from ``can_be_set`` so they can be
+        returned later.
+    """
+
+    def __init__(
+        self,
+        event_name: str,
+        check: CheckFunction
+    ):
+        self.event_name = event_name
+        self.check = check
+        self.event = Event()
+        self.return_value = None
+        super().__init__()
+
+    async def wait(self):
+        """
+        Waits until ``self.event`` is set.
+        """
+        await self.event.wait()
+
+    def process(self, event_name: str, *args) -> bool:
+        if self.matches_event(event_name, *args):
+            self.return_value = args
+            self.event.set()
+
+
+class _LoopEmptyError(Exception):
+    "Raised when the _LoopMgr is empty and cannot accept new item"
+
+
+class _LoopMgr(_Processable):
+    """
+    Parameters
+    ----------
+    event_name : str
+        The name of the event.
+    check : Optional[Callable[[Any], bool]]
+        ``can_be_set`` only returns true if this function returns true.
+        Will be ignored if set to None.
+
+    Attributes
+    ----------
+    can_expand : bool
+        Whether the queue is allowed to grow. Turned to false once the
+        EventMgr's timer runs out.
+    events : :class:`collections.deque`
+        Qeue of events to be processed.
+    wait : :class:`asyncio.Event`
+        Used to make ``get_next()` wait for the next event.
+    """
+
+    def __init__(self, event_name: str, check: CheckFunction) -> None:
+        self.event_name = event_name
+        self.check = check
+
+        self.can_expand = True
+        self.events = deque()
+        self.wait = Event()
+
+    def process(self, event_name: str, *args):
+        if not self.can_expand:
+            return
+
+        if self.matches_event(event_name, *args):
+            self.events.append(args)
+            self.wait.set()
+
+    async def get_next(self):
+        """
+        Returns the next item if the queue. If there are no items in the queue,
+        it will return the next event that happens.
+        """
+        if len(self.events) == 0:
+            if not self.can_expand:
+                raise _LoopEmptyError
+
+            self.wait.clear()
+            await self.wait.wait()
+            return self.events.popleft()
+        else:
+            return self.events.popleft()
+
+
+class EventMgr:
+    """
+    Attributes
+    ----------
+    event_list : List[_DiscordEvent]
+        The List of events that need to be processed.
+    """
+
+    def __init__(self):
+        self.event_list: _Processable = []
```

Comment:
Is that definitely the right typehint?

Reply:
Im sure this is the right one. `event_list` stores the `_LoopMgr` class and `_Event` class since the client can run the exact same functions on both of them.

Reply:
What I meant was why is this typed as `_Processable` and then set to a list. Did you mean `list[_Processable]`?